File: src/main/scala/msgpack4z/Code.java
package msgpack4z;

final class Code {
    private Code() {
        throw new RuntimeException("should not create instance");
    }

    public static boolean isFixInt(byte b) {
        final int v = b & 0xFF;
        return v <= 0x7f || v >= 0xe0;
    }

    public static boolean isPosFixInt(byte b) {
        return (b & POSFIXINT_MASK) == 0;
    }

    public static boolean isNegFixInt(byte b) {
        return (b & NEGFIXINT_PREFIX) == NEGFIXINT_PREFIX;
    }

    public static boolean isFixStr(byte b) {
        return (b & (byte) 0xe0) == Code.FIXSTR_PREFIX;
    }

    public static boolean isFixedArray(byte b) {
        return (b & (byte) 0xf0) == Code.FIXARRAY_PREFIX;
    }

    public static boolean isFixedMap(byte b) {
        return (b & (byte) 0xe0) == Code.FIXMAP_PREFIX;
    }

    public static boolean isFixedRaw(byte b) {
        return (b & (byte) 0xe0) == Code.FIXSTR_PREFIX;
    }

    public static final byte POSFIXINT_MASK = (byte) 0x80;

    public static final byte FIXMAP_PREFIX = (byte) 0x80;
    public static final byte FIXARRAY_PREFIX = (byte) 0x90;
    public static final byte FIXSTR_PREFIX = (byte) 0xa0;

    public static final byte NIL = (byte) 0xc0;
    public static final byte FALSE = (byte) 0xc2;
    public static final byte TRUE = (byte) 0xc3;
    public static final byte BIN8 = (byte) 0xc4;
    public static final byte BIN16 = (byte) 0xc5;
    public static final byte BIN32 = (byte) 0xc6;
    public static final byte EXT8 = (byte) 0xc7;
    public static final byte EXT16 = (byte) 0xc8;
    public static final byte EXT32 = (byte) 0xc9;
    public static final byte FLOAT32 = (byte) 0xca;
    public static final byte FLOAT64 = (byte) 0xcb;
    public static final byte UINT8 = (byte) 0xcc;
    public static final byte UINT16 = (byte) 0xcd;
    public static final byte UINT32 = (byte) 0xce;
    public static final byte UINT64 = (byte) 0xcf;

    public static final byte INT8 = (byte) 0xd0;
    public static final byte INT16 = (byte) 0xd1;
    public static final byte INT32 = (byte) 0xd2;
    public static final byte INT64 = (byte) 0xd3;

    public static final byte FIXEXT1 = (byte) 0xd4;
    public static final byte FIXEXT2 = (byte) 0xd5;
    public static final byte FIXEXT4 = (byte) 0xd6;
    public static final byte FIXEXT8 = (byte) 0xd7;
    public static final byte FIXEXT16 = (byte) 0xd8;

    public static final byte STR8 = (byte) 0xd9;
    public static final byte STR16 = (byte) 0xda;
    public static final byte STR32 = (byte) 0xdb;

    public static final byte ARRAY16 = (byte) 0xdc;
    public static final byte ARRAY32 = (byte) 0xdd;

    public static final byte MAP16 = (byte) 0xde;
    public static final byte MAP32 = (byte) 0xdf;

    public static final byte NEGFIXINT_PREFIX = (byte) 0xe0;

    public static MsgType getType(final byte b) {
        final MsgType t = formatTable[b & 0xff];
        if (t == null) {
            throw new RuntimeException("invalid byte " + b);
        } else {
            return t;
        }
    }

    private final static MsgType[] formatTable = new MsgType[0x100];

    static {
        for (int b = 0; b <= 0xFF; ++b) {
            formatTable[b] = getType0((byte) b);
        }
    }

    private static MsgType getType0(final byte b) {
        if (isPosFixInt(b) || isNegFixInt(b)) {
            return MsgType.INTEGER;
        }
        if (Code.isFixStr(b)) {
            return MsgType.STRING;
        }
        if (Code.isFixedArray(b)) {
            return MsgType.ARRAY;
        }
        if (Code.isFixedMap(b)) {
            return MsgType.MAP;
        }
        switch (b) {
            case Code.NIL:
                return MsgType.NIL;
            case Code.FALSE:
            case Code.TRUE:
                return MsgType.BOOLEAN;
            case Code.BIN8:
            case Code.BIN16:
            case Code.BIN32:
                return MsgType.BINARY;
            case Code.FLOAT32:
            case Code.FLOAT64:
                return MsgType.FLOAT;
            case Code.UINT8:
            case Code.UINT16:
            case Code.UINT32:
            case Code.UINT64:
            case Code.INT8:
            case Code.INT16:
            case Code.INT32:
            case Code.INT64:
                return MsgType.INTEGER;
            case Code.STR8:
            case Code.STR16:
            case Code.STR32:
                return MsgType.STRING;
            case Code.ARRAY16:
            case Code.ARRAY32:
                return MsgType.ARRAY;
            case Code.MAP16:
            case Code.MAP32:
                return MsgType.MAP;
            case Code.FIXEXT1:
            case Code.FIXEXT2:
            case Code.FIXEXT4:
            case Code.FIXEXT8:
            case Code.FIXEXT16:
            case Code.EXT8:
            case Code.EXT16:
            case Code.EXT32:
                return MsgType.EXTENDED;

            default:
                return null;
        }
    }
}

File: src/main/scala/msgpack4z/Code.scala
package msgpack4z

object Code {
  def isFixInt(b: Byte): Boolean = {
    val v: Int = b & 0xFF
    v <= 0x7f || v >= 0xe0
  }

  def isPosFixInt(b: Byte): Boolean = {
    (b & POSFIXINT_MASK) == 0
  }

  def isNegFixInt(b: Byte): Boolean = {
    (b & NEGFIXINT_PREFIX) == NEGFIXINT_PREFIX
  }

  def isFixStr(b: Byte): Boolean = {
    (b & 0xe0.toByte) == Code.FIXSTR_PREFIX
  }

  def isFixedArray(b: Byte): Boolean = {
    (b & 0xf0.toByte) == Code.FIXARRAY_PREFIX
  }

  def isFixedMap(b: Byte): Boolean = {
    (b & 0xf0.toByte) == Code.FIXMAP_PREFIX
  }

  def isFixedRaw(b: Byte): Boolean = {
    (b & 0xe0.toByte) == Code.FIXSTR_PREFIX
  }

  val POSFIXINT_MASK: Byte = 0x80.toByte
  val FIXMAP_PREFIX: Byte = 0x80.toByte
  val FIXARRAY_PREFIX: Byte = 0x90.toByte
  val FIXSTR_PREFIX: Byte = 0xa0.toByte
  val NIL: Byte = 0xc0.toByte
  val FALSE: Byte = 0xc2.toByte
  val TRUE: Byte = 0xc3.toByte
  val BIN8: Byte = 0xc4.toByte
  val BIN16: Byte = 0xc5.toByte
  val BIN32: Byte = 0xc6.toByte
  val EXT8: Byte = 0xc7.toByte
  val EXT16: Byte = 0xc8.toByte
  val EXT32: Byte = 0xc9.toByte
  val FLOAT32: Byte = 0xca.toByte
  val FLOAT64: Byte = 0xcb.toByte
  val UINT8: Byte = 0xcc.toByte
  val UINT16: Byte = 0xcd.toByte
  val UINT32: Byte = 0xce.toByte
  val UINT64: Byte = 0xcf.toByte
  val INT8: Byte = 0xd0.toByte
  val INT16: Byte = 0xd1.toByte
  val INT32: Byte = 0xd2.toByte
  val INT64: Byte = 0xd3.toByte
  val FIXEXT1: Byte = 0xd4.toByte
  val FIXEXT2: Byte = 0xd5.toByte
  val FIXEXT4: Byte = 0xd6.toByte
  val FIXEXT8: Byte = 0xd7.toByte
  val FIXEXT16: Byte = 0xd8.toByte
  val STR8: Byte = 0xd9.toByte
  val STR16: Byte = 0xda.toByte
  val STR32: Byte = 0xdb.toByte
  val ARRAY16: Byte = 0xdc.toByte
  val ARRAY32: Byte = 0xdd.toByte
  val MAP16: Byte = 0xde.toByte
  val MAP32: Byte = 0xdf.toByte
  val NEGFIXINT_PREFIX: Byte = 0xe0.toByte

  def getType(b: Byte): MsgType = {
    val t = formatTable(b & 0xff)
    if (t == null) {
      throw new RuntimeException("invalid byte " + b)
    } else {
      t
    }
  }

  private[this] val formatTable: Array[MsgType] = {
    val array = new Array[MsgType](0x100)
    var b = 0
    while (b <= 0xFF) {
      array(b) = getType0(b.toByte)
      b += 1
    }
    array
  }

  private[this] def getType0(b: Byte): MsgType = {
    if (isPosFixInt(b) || isNegFixInt(b)) {
      MsgType.INTEGER
    } else if (Code.isFixStr(b)) {
      MsgType.STRING
    } else if (Code.isFixedArray(b)) {
      MsgType.ARRAY
    } else if (Code.isFixedMap(b)) {
      MsgType.MAP
    } else {
      b match {
        case Code.NIL =>
          MsgType.NIL
        case Code.FALSE | Code.TRUE =>
          MsgType.BOOLEAN
        case Code.BIN8 | Code.BIN16 | Code.BIN32 =>
          MsgType.BINARY
        case Code.FLOAT32 | Code.FLOAT64 =>
          MsgType.FLOAT
        case Code.UINT8 | Code.UINT16 | Code.UINT32 | Code.UINT64 | Code.INT8 | Code.INT16 | Code.INT32 | Code.INT64 =>
          MsgType.INTEGER
        case Code.STR8 | Code.STR16 | Code.STR32 =>
          MsgType.STRING
        case Code.ARRAY16 | Code.ARRAY32 =>
          MsgType.ARRAY
        case Code.MAP16 | Code.MAP32 =>
          MsgType.MAP
        case Code.FIXEXT1 | Code.FIXEXT2 | Code.FIXEXT4 | Code.FIXEXT8 | Code.FIXEXT16 | Code.EXT8 | Code.EXT16 | Code.EXT32 =>
          MsgType.EXTENSION
        case _ =>
          null
      }
    }
  }

}
